packages/product/src/services/product-module-service.ts

```diff
@@ -369,6 +375,145 @@ export default class ProductModuleService<
     return JSON.parse(JSON.stringify(products))
   }
 
+  // TODO: cleanup, abstract common create and update steps
+  async update(
+    data: ProductTypes.UpdateProductDTO[],
+    sharedContext?: Context
+  ): Promise<ProductTypes.ProductDTO[]> {
+    const products = await this.baseRepository_.transaction(
+      async (manager) => {
+        sharedContext = sharedContext || {
+          transactionManager: manager,
+        }
+
+        const productVariantsMap = new Map<
+          string,
+          ProductTypes.CreateProductVariantDTO[]
+        >()
+        const productOptionsMap = new Map<
+          string,
+          ProductTypes.CreateProductOptionDTO[]
+        >()
+
+        const productsData = await Promise.all(
+          data.map(async (product) => {
+            const productData = { ...product }
+
+            if (!isDefined(productData.id)) {
+              throw new MedusaError(
+                MedusaError.Types.NOT_FOUND,
+                `Cannot update product without id`
+              )
+            }
+
+            const variants = productData.variants
+            const options = productData.options
+            delete productData.options
+            delete productData.variants
+
+            productVariantsMap.set(productData.id, variants ?? [])
+            productOptionsMap.set(productData.id, options ?? [])
+
+            if (!productData.thumbnail && productData.images?.length) {
+              productData.thumbnail = isString(productData.images[0])
+                ? (productData.images[0] as string)
+                : (productData.images[0] as { url: string }).url
+            }
+
+            if (productData.is_giftcard) {
+              productData.discountable = false
+            }
+
+            if (productData.images?.length) {
+              productData.images = await this.productImageService_.upsert(
+                productData.images.map((image) =>
+                  isString(image) ? image : image.url
+                ),
+                sharedContext
+              )
+            }
+
+            if (productData.tags?.length) {
+              productData.tags = await this.productTagService_.upsert(
+                productData.tags,
+                sharedContext
+              )
+            }
+
+            if (isDefined(productData.type)) {
+              const productType = (
+                await this.productTypeService_.upsert(
+                  [productData.type as ProductTypes.CreateProductTypeDTO],
+                  sharedContext
+                )
+              )
+
+              productData.type = productType?.[0]
+            }
+
+            return productData as ProductServiceTypes.UpdateProductDTO
+          })
+        )
+
+        // TODO
+        // Shipping profile is not part of the module
+        // as well as sales channel
```

Comment:
todo: I think we can now remove those todo's they are part of the workflow, you can remove it from the create as well